lib/rex/proto/ldap/server.rb

```diff
@@ -149,76 +156,126 @@ def dispatch_request(cli, data)
+                      result_code,
+                      '',
+                      Net::LDAP::ResultStrings[result_code],
+                      Net::LDAP::PDU::SearchResult
+                    )
+                  when Net::LDAP::PDU::UnbindRequest
+                    client.close
+                    nil
+                  else
+                    if suitable_response(pdu.app_tag)
+                      result_code = Net::LDAP::ResultCodeUnwillingToPerform
+                      encode_ldap_response(
+                        pdu.message_id,
+                        result_code,
+                        '',
+                        Net::LDAP::ResultStrings[result_code],
+                        suitable_response(pdu.app_tag)
+                      )
+                    else
+                      client.close
+                    end
+                  end
+
+            if @pdu_process && post_pdu
+              @pdu_process.call(processed_pdu_data)
+            end
+            send_response(client, res) unless res.nil?
           rescue StandardError => e
-            elog(e)
-            cli.close
-            raise e
+            client.close
           end
         end
 
         #
         # Encode response for LDAP client consumption
         #
-        # @param msgid [Integer] LDAP message identifier
-        # @param code  [Integer] LDAP message code
-        # @param dn    [String]  LDAP distinguished name
-        # @param msg   [String]  LDAP response message
-        # @param tag   [Integer] LDAP response tag
+        # @param msgid        [Integer] LDAP message identifier
+        # @param code         [Integer] LDAP message code
+        # @param dn           [String]  LDAP distinguished name
+        # @param msg          [String]  LDAP response message
+        # @param tag          [Integer] LDAP response tag
+        # @param server_creds [String]  LDAP response server SASL credentials
         #
         # @return [Net::BER::BerIdentifiedOid] LDAP query response
-        def encode_ldap_response(msgid, code, dn, msg, tag)
-          [
-            msgid.to_ber,
+        def encode_ldap_response(msgid, code, dn, msg, tag, server_creds = '')
+          case code
+          when Net::LDAP::ResultCodeSaslBindInProgress
             [
-              code.to_ber_enumerated,
-              dn.to_ber,
-              msg.to_ber
-            ].to_ber_appsequence(tag)
-          ].to_ber_sequence
+              msgid.to_ber,
+              [
+                code.to_ber_enumerated,
+                dn.to_ber,
+                msg.to_ber,
+                [server_creds.to_ber].to_ber_contextspecific(7)
+              ].to_ber_appsequence(tag)
+            ].to_ber_sequence
+          else
+            [
+              msgid.to_ber,
+              [
+                code.to_ber_enumerated,
+                dn.to_ber,
+                msg.to_ber
+              ].to_ber_appsequence(tag)
+            ].to_ber_sequence
+          end
```

Comment:
Can this be more generic? Consider:

``` Ruby
        #
        # Encode response for LDAP client consumption
        #
        # @param msgid      [Integer] LDAP message identifier
        # @param code       [Integer] LDAP message code
        # @param dn         [String]  LDAP distinguished name
        # @param msg        [String]  LDAP response message
        # @param tag        [Integer] LDAP response tag
        # @param extra_data [Hash]    Hash containtining `data` and `context_code` keys when required
        #
        # @return [Net::BER::BerIdentifiedOid] LDAP query response
        def encode_ldap_response(msgid, code, dn, msg, tag, extra_data = nil)
          tag_sequence = [
            code.to_ber_enumerated,
            dn.to_ber,
            msg.to_ber
          ]

          if extra_data && extra_data[:data] && extra_data[:context_code]
            tag_sequence << [extra_data[:data].to_ber].to_ber_contextspecific(extra_data[:context_code])
          end

          [
            msgid.to_ber,
            tag_sequence.to_ber_appsequence(tag)
          ].to_ber_sequence
        end
```

This idea is untested, I believe this would be more flexible if other LDAP responses can contain an extra object like `Net::LDAP::ResultCodeSaslBindInProgress` requires.

Reply:
This could method could alternately be adjusted to avoid a hash since it only adds two parameters as:

``` Ruby
        #
        # Encode response for LDAP client consumption
        #
        # @param msgid        [Integer] LDAP message identifier
        # @param code         [Integer] LDAP message code
        # @param dn           [String]  LDAP distinguished name
        # @param msg          [String]  LDAP response message
        # @param tag          [Integer] LDAP response tag
        # @param context_data [String]  Additional data to serialize in the sequence
        # @param context_code [Integer] Context Specific code releated to `context_data`
        #
        # @return [Net::BER::BerIdentifiedOid] LDAP query response
        def encode_ldap_response(msgid, code, dn, msg, tag, context_data: nil, context_code: nil)
          tag_sequence = [
            code.to_ber_enumerated,
            dn.to_ber,
            msg.to_ber
          ]

          if context_data && context_code
            tag_sequence << [context_data.to_ber].to_ber_contextspecific(context_code)
          end

          [
            msgid.to_ber,
            tag_sequence.to_ber_appsequence(tag)
          ].to_ber_sequence
        end
```

Reply:
Thanks